File: nmos/raft/operations.py
# Copyright (C) 2025-2026 Alain Bouchard
# SPDX-License-Identifier: Apache-2.0

"""The replicated unit: a registry operation, not a key-value pair.

This is the design decision the whole package turns on. The etcd backend
replicates *writes to a key-value store* and rebuilds a registry view from
them, which is why it needs an envelope format, a key layout, a watch, and a
fence to tell it when its view has caught up to a write it just made. Here the
log carries the operation itself, so applying a committed entry **is** the
store mutation plus its grain publication, and none of that machinery exists.

Determinism is the whole contract
---------------------------------
Every member applies the same entries in the same order and must reach the same
state. That makes the choice of what travels in an operation a correctness
question rather than an efficiency one, and it produces one rule:

    **Anything a member would otherwise read from its local environment must
    be decided once, by the proposer, and carried.**

Concretely, ``created``/``updated`` cursors and ``health`` are fields here
rather than defaults filled in at apply time, because ``apply_committed`` would
otherwise call ``health_now()`` and ``_next_cursor()`` -- both of which read
local state and would give a different answer on every member. Those two
defaults are precisely the ones ``machine.py`` must always override.

What is *not* carried, and why
------------------------------
The proposer's ``PreparedRegistration`` is not on the wire. It is tempting --
validation has already happened, so why do it twice -- but it would be wrong:
``store.prepare`` decides five things, and four of them are scoped to the
Node's own subtree while one, the id-uniqueness check against ``_type_of``, is
global across every Node. An owner cannot decide that one, because a
concurrent registration under a different Node on a different member might be
claiming the same id.

The log is what serialises those, so apply re-runs ``prepare`` against the
replicated store and *that* answer is authoritative. What travels instead is
``expect_created``: the proposer's belief about 201-vs-200, carried purely as a
tripwire. If apply disagrees, the two members have diverged, and saying so
loudly beats serving two different answers quietly.

Bodies travel verbatim
----------------------
``body_text`` is the request body exactly as the client sent it. The registry's
guarantee is that what was registered is what every member serves, byte for
byte, including vendor extensions the generated types do not model -- so a
re-encode here, however lossless it looked, would break that guarantee at the
one point nobody would think to check.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
from typing import Union

from nmos.raft.errors import RaftProtocolError
from nmos.raft.wire import Reader, Writer
from nmos.registry.types import ResourceType, TaiCursor


class OpKind(IntEnum):
    """What an entry does. Numbers are permanent, like every wire number."""

    NOOP = 0
    """A leader's term-establishing entry.

    Raft §8: a new leader cannot know what is committed from previous terms
    until it commits an entry of its own, so it appends one that does nothing.
    """

    REGISTER = 1
    UNREGISTER = 2

    EXPIRE = 3
    """A Node whose lease lapsed, and its whole subtree.

    Decided by the owner against its own clock and then replicated, rather than
    each member expiring on its own clock -- which is how members end up
    disagreeing about which Nodes are alive.
    """

    FORGET = 4
    """Drop tombstones whose forget interval has elapsed.

    Replicated for the same reason and with the same shape as EXPIRE: the
    decision reads a clock, so it is made once and the *result* is what every
    member applies.
    """

    CLAIM_OWNERSHIP = 5
    RELEASE_OWNERSHIP = 6

    MEMBER_DOWN = 7
    """One member is gone; release every Node it owned, atomically.

    A single entry rather than one per Node, so a member holding a thousand
    Nodes does not put a thousand entries through consensus at the moment the
    cluster is already one member down.
    """


@dataclass(frozen=True)
class ProposalId:
    """Identifies a proposal so its originator can be answered.

    ``(member, sequence)`` rather than a UUID: it is two varints instead of
    sixteen bytes on every entry, and it is ordered.

    The sequence must be unique over the member's whole **history**, not just
    over one run of it. A restarted member has no outstanding waiters of its
    own, which is what an earlier version of this docstring relied on -- but
    its *entries* survive it, sitting in the cluster's log and applying after
    it returns. If the new incarnation mints the same ids, an old entry's
    outcome resolves a new caller's future: a registration answered with an
    unregistration's result. ``node.py`` therefore seeds the sequence from the
    incarnation, giving each run of the member its own range.
    """

    member: int
    sequence: int


@dataclass(frozen=True)
class NoopOp:
    KIND = OpKind.NOOP
    proposal: ProposalId


@dataclass(frozen=True)
class RegisterOp:
    """Register or update one resource."""

    KIND = OpKind.REGISTER

    proposal: ProposalId
    resource_type: ResourceType
    resource_id: str
    node_id: str
    body_text: str
    created: TaiCursor
    updated: TaiCursor
    health: int
    expect_created: bool
    claim_owner: int | None = None
    """Fused ownership claim.

    When the proposer is taking ownership of a previously unowned Node, the
    claim rides along on the registration instead of being a separate entry.
    That keeps a Node's first registration to one round trip rather than two,
    which matters because a facility powering up is entirely first
    registrations.
    """


@dataclass(frozen=True)
class UnregisterOp:
    KIND = OpKind.UNREGISTER

    proposal: ProposalId
    resource_type: ResourceType
    resource_id: str


@dataclass(frozen=True)
class ExpireOp:
    KIND = OpKind.EXPIRE

    proposal: ProposalId
    node_id: str


@dataclass(frozen=True)
class ForgetOp:
    KIND = OpKind.FORGET

    proposal: ProposalId
    victims: tuple[tuple[ResourceType, str], ...]


@dataclass(frozen=True)
class ClaimOwnershipOp:
    KIND = OpKind.CLAIM_OWNERSHIP

    proposal: ProposalId
    node_id: str
    owner: int


@dataclass(frozen=True)
class ReleaseOwnershipOp:
    KIND = OpKind.RELEASE_OWNERSHIP

    proposal: ProposalId
    node_id: str


@dataclass(frozen=True)
class MemberDownOp:
    KIND = OpKind.MEMBER_DOWN

    proposal: ProposalId
    member: int


RegistryOperation = Union[
    NoopOp,
    RegisterOp,
    UnregisterOp,
    ExpireOp,
    ForgetOp,
    ClaimOwnershipOp,
    ReleaseOwnershipOp,
    MemberDownOp,
]


# ---------------------------------------------------------------------------
# Codec
# ---------------------------------------------------------------------------

def _write_proposal(writer: Writer, proposal: ProposalId) -> None:
    writer.bytes_(
        1,
        Writer().uint(1, proposal.member).uint(2, proposal.sequence).take(),
    )


def _read_proposal(payload: bytes) -> ProposalId:
    member = sequence = 0
    reader = Reader(payload)
    for number, wire in reader:
        if number == 1:
            member = reader.uint()
        elif number == 2:
            sequence = reader.uint()
        else:
            reader.skip(wire)
    return ProposalId(member=member, sequence=sequence)


def _write_cursor(writer: Writer, field: int, cursor: TaiCursor) -> None:
    writer.bytes_(
        field,
        Writer().uint(1, cursor.seconds).uint(2, cursor.nanoseconds).take(),
    )


def _read_cursor(payload: bytes) -> TaiCursor:
    seconds = nanoseconds = 0
    reader = Reader(payload)
    for number, wire in reader:
        if number == 1:
            seconds = reader.uint()
        elif number == 2:
            nanoseconds = reader.uint()
        else:
            reader.skip(wire)
    return TaiCursor(seconds, nanoseconds)


def _resource_type(value: str) -> ResourceType:
    try:
        return ResourceType(value)
    except ValueError as exc:
        raise RaftProtocolError(f"unknown resource type {value!r}") from exc


def encode_operation(op: RegistryOperation) -> bytes:
    """Serialise one operation. Field 1 is always the kind."""
    writer = Writer().uint(1, int(op.KIND))
    body = Writer()
    _write_proposal(body, op.proposal)

    if isinstance(op, RegisterOp):
        body.string(2, op.resource_type.value)
        body.string(3, op.resource_id)
        body.string(4, op.node_id)
        body.string(5, op.body_text)
        _write_cursor(body, 6, op.created)
        _write_cursor(body, 7, op.updated)
        body.uint(8, op.health)
        body.bool_(9, op.expect_created)
        # Present-or-absent: member 0 is a real member, so a sentinel zero
        # would claim ownership for the first member on every registration
        # that was not claiming anything.
        if op.claim_owner is not None:
            body.uint(10, op.claim_owner)
    elif isinstance(op, UnregisterOp):
        body.string(2, op.resource_type.value)
        body.string(3, op.resource_id)
    elif isinstance(op, ExpireOp):
        body.string(2, op.node_id)
    elif isinstance(op, ForgetOp):
        for resource_type, resource_id in op.victims:
            body.bytes_(
                2,
                Writer().string(1, resource_type.value)
                .string(2, resource_id).take(),
            )
    elif isinstance(op, ClaimOwnershipOp):
        body.string(2, op.node_id)
        body.uint(3, op.owner)
    elif isinstance(op, ReleaseOwnershipOp):
        body.string(2, op.node_id)
    elif isinstance(op, MemberDownOp):
        body.uint(2, op.member)

    writer.bytes_(2, body.take())
    return writer.take()


def decode_operation(data: bytes) -> RegistryOperation:
    """Parse one operation, or raise :class:`RaftProtocolError`.

    Called on receipt rather than at apply time, deliberately. A malformed
    entry discovered here can still drop the link; discovered inside apply it
    would be a synchronous mutation that has nowhere to fail.
    """
    kind_value: int | None = None
    body = b""
    reader = Reader(data)
    for number, wire in reader:
        if number == 1:
            kind_value = reader.uint()
        elif number == 2:
            body = reader.bytes_()
        else:
            reader.skip(wire)

    if kind_value is None:
        raise RaftProtocolError("operation has no kind")
    try:
        kind = OpKind(kind_value)
    except ValueError as exc:
        raise RaftProtocolError(f"unknown operation kind {kind_value}") from exc

    proposal = ProposalId(member=0, sequence=0)
    resource_type_name = ""
    resource_id = node_id = body_text = ""
    created = updated = TaiCursor.min()
    health = 0
    expect_created = False
    claim_owner: int | None = None
    member = 0
    victims: list[tuple[ResourceType, str]] = []

    inner = Reader(body)
    for number, wire in inner:
        if number == 1:
            proposal = _read_proposal(inner.bytes_())
        elif kind is OpKind.REGISTER:
            if number == 2:
                resource_type_name = inner.string()
            elif number == 3:
                resource_id = inner.string()
            elif number == 4:
                node_id = inner.string()
            elif number == 5:
                body_text = inner.string()
            elif number == 6:
                created = _read_cursor(inner.bytes_())
            elif number == 7:
                updated = _read_cursor(inner.bytes_())
            elif number == 8:
                health = inner.uint()
            elif number == 9:
                expect_created = inner.bool_()
            elif number == 10:
                claim_owner = inner.uint()
            else:
                inner.skip(wire)
        elif kind is OpKind.UNREGISTER:
            if number == 2:
                resource_type_name = inner.string()
            elif number == 3:
                resource_id = inner.string()
            else:
                inner.skip(wire)
        elif kind in (OpKind.EXPIRE, OpKind.RELEASE_OWNERSHIP):
            if number == 2:
                node_id = inner.string()
            else:
                inner.skip(wire)
        elif kind is OpKind.FORGET:
            if number == 2:
                victims.append(_read_victim(inner.bytes_()))
            else:
                inner.skip(wire)
        elif kind is OpKind.CLAIM_OWNERSHIP:
            if number == 2:
                node_id = inner.string()
            elif number == 3:
                claim_owner = inner.uint()
            else:
                inner.skip(wire)
        elif kind is OpKind.MEMBER_DOWN:
            if number == 2:
                member = inner.uint()
            else:
                inner.skip(wire)
        else:
            inner.skip(wire)

    if kind is OpKind.NOOP:
        return NoopOp(proposal=proposal)
    if kind is OpKind.REGISTER:
        return RegisterOp(
            proposal=proposal,
            resource_type=_resource_type(resource_type_name),
            resource_id=resource_id, node_id=node_id, body_text=body_text,
            created=created, updated=updated, health=health,
            expect_created=expect_created, claim_owner=claim_owner,
        )
    if kind is OpKind.UNREGISTER:
        return UnregisterOp(
            proposal=proposal,
            resource_type=_resource_type(resource_type_name),
            resource_id=resource_id,
        )
    if kind is OpKind.EXPIRE:
        return ExpireOp(proposal=proposal, node_id=node_id)
    if kind is OpKind.FORGET:
        return ForgetOp(proposal=proposal, victims=tuple(victims))
    if kind is OpKind.CLAIM_OWNERSHIP:
        if claim_owner is None:
            raise RaftProtocolError("ownership claim names no owner")
        return ClaimOwnershipOp(
            proposal=proposal, node_id=node_id, owner=claim_owner,
        )
    if kind is OpKind.RELEASE_OWNERSHIP:
        return ReleaseOwnershipOp(proposal=proposal, node_id=node_id)
    return MemberDownOp(proposal=proposal, member=member)


def _read_victim(payload: bytes) -> tuple[ResourceType, str]:
    name = resource_id = ""
    reader = Reader(payload)
    for number, wire in reader:
        if number == 1:
            name = reader.string()
        elif number == 2:
            resource_id = reader.string()
        else:
            reader.skip(wire)
    return _resource_type(name), resource_id

File: rust/crates/nmos-registry-raft/src/operations.rs
// Copyright (C) 2025-2026 Alain Bouchard
// SPDX-License-Identifier: Apache-2.0

//! The replicated unit: a registry operation, not a key-value pair.
//!
//! Port of `nmos/raft/operations.py`.
//!
//! This is the design decision the whole package turns on. The etcd backend
//! replicates *writes to a key-value store* and rebuilds a registry view from
//! them, which is why it needs an envelope format, a key layout, a watch, and a
//! fence to tell it when its view has caught up to a write it just made. Here
//! the log carries the operation itself, so applying a committed entry **is**
//! the store mutation plus its grain publication, and none of that machinery
//! exists.
//!
//! # Determinism is the whole contract
//!
//! Every member applies the same entries in the same order and must reach the
//! same state. That makes the choice of what travels in an operation a
//! correctness question rather than an efficiency one, and it produces one
//! rule:
//!
//! > **Anything a member would otherwise read from its local environment must
//! > be decided once, by the proposer, and carried.**
//!
//! Concretely, `created`/`updated` cursors and `health` are fields here rather
//! than defaults filled in at apply time, because the store's apply would
//! otherwise read the clock and the local cursor allocator -- both of which
//! give a different answer on every member. Those two defaults are precisely
//! the ones the state machine must always override.
//!
//! # What is *not* carried, and why
//!
//! The proposer's prepared registration is not on the wire. It is tempting --
//! validation has already happened, so why do it twice -- but it would be
//! wrong: `store.prepare` decides five things, and four of them are scoped to
//! the Node's own subtree while one, the id-uniqueness check, is global across
//! every Node. An owner cannot decide that one, because a concurrent
//! registration under a different Node on a different member might be claiming
//! the same id.
//!
//! The log is what serialises those, so apply re-runs `prepare` against the
//! replicated store and *that* answer is authoritative. What travels instead is
//! `expect_created`: the proposer's belief about 201-vs-200, carried purely as
//! a tripwire. If apply disagrees, the two members have diverged, and saying so
//! loudly beats serving two different answers quietly.
//!
//! # Bodies travel verbatim
//!
//! `body_text` is the request body exactly as the client sent it. The
//! registry's guarantee is that what was registered is what every member
//! serves, byte for byte, including vendor extensions the generated types do
//! not model -- so a re-encode here, however lossless it looked, would break
//! that guarantee at the one point nobody would think to check.
//!
//! # Shape: one struct, one kind
//!
//! The Python gives each operation its own frozen dataclass, each carrying its
//! own `proposal`. Here the proposal is lifted out into [`Operation`] and the
//! rest is an [`OperationKind`] enum, which is what the encoding already does
//! -- field 1 of the body is the proposal, uniformly, for every kind. Lifting
//! it makes that structural rather than repeated, and makes "every operation
//! answers a proposal" impossible to forget when a kind is added.

use nmos_registry_core::cursor::TaiCursor;
use nmos_registry_core::resource_type::ResourceType;

use crate::errors::RaftProtocolError;
use crate::wire::{Reader, Writer};

/// What an entry does. Numbers are permanent, like every wire number.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum OpKind {
    /// A leader's term-establishing entry.
    ///
    /// Raft §8: a new leader cannot know what is committed from previous terms
    /// until it commits an entry of its own, so it appends one that does
    /// nothing.
    Noop = 0,
    /// Register or update one resource.
    Register = 1,
    /// Remove one resource and its subtree.
    Unregister = 2,
    /// A Node whose lease lapsed, and its whole subtree.
    ///
    /// Decided by the owner against its own clock and then replicated, rather
    /// than each member expiring on its own clock -- which is how members end
    /// up disagreeing about which Nodes are alive.
    Expire = 3,
    /// Drop tombstones whose forget interval has elapsed.
    ///
    /// Replicated for the same reason and with the same shape as
    /// [`Self::Expire`]: the decision reads a clock, so it is made once and the
    /// *result* is what every member applies.
    Forget = 4,
    /// Take responsibility for a Node's subtree.
    ClaimOwnership = 5,
    /// Give it up.
    ReleaseOwnership = 6,
    /// One member is gone; release every Node it owned, atomically.
    ///
    /// A single entry rather than one per Node, so a member holding a thousand
    /// Nodes does not put a thousand entries through consensus at the moment
    /// the cluster is already one member down.
    MemberDown = 7,
}

impl OpKind {
    const fn from_byte(value: u64) -> Option<Self> {
        match value {
            0 => Some(Self::Noop),
            1 => Some(Self::Register),
            2 => Some(Self::Unregister),
            3 => Some(Self::Expire),
            4 => Some(Self::Forget),
            5 => Some(Self::ClaimOwnership),
            6 => Some(Self::ReleaseOwnership),
            7 => Some(Self::MemberDown),
            _ => None,
        }
    }
}

/// Identifies a proposal so its originator can be answered.
///
/// `(member, sequence)` rather than a UUID: it is two varints instead of
/// sixteen bytes on every entry, and it is ordered.
///
/// The sequence must be unique over the member's whole **history**, not just
/// over one run of it. A restarted member has no outstanding waiters of its
/// own -- but its *entries* survive it, sitting in the cluster's log and
/// applying after it returns. If the new incarnation mints the same ids, an old
/// entry's outcome resolves a new caller's future: a registration answered with
/// an unregistration's result. The node therefore seeds the sequence from the
/// incarnation, giving each run of the member its own range.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ProposalId {
    /// Which member proposed it.
    pub member: u64,
    /// Unique within that member's whole history. See the struct docs.
    pub sequence: u64,
}

/// Register or update one resource.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Register {
    /// Which collection it belongs to.
    pub resource_type: ResourceType,
    /// The resource's own id.
    pub resource_id: String,
    /// The Node at the root of its subtree, for ownership.
    pub node_id: String,
    /// The request body, verbatim. See the module docs.
    pub body_text: String,
    /// The creation cursor, decided by the proposer.
    pub created: TaiCursor,
    /// The update cursor, decided by the proposer.
    pub updated: TaiCursor,
    /// The health instant, decided by the proposer.
    pub health: u64,
    /// The proposer's belief about 201-versus-200, carried as a tripwire.
    pub expect_created: bool,
    /// Fused ownership claim.
    ///
    /// When the proposer is taking ownership of a previously unowned Node, the
    /// claim rides along on the registration instead of being a separate entry.
    /// That keeps a Node's first registration to one round trip rather than
    /// two, which matters because a facility powering up is entirely first
    /// registrations.
    ///
    /// `None` rather than zero: member 0 is a real member, so a sentinel would
    /// claim ownership for the first member on every registration that was not
    /// claiming anything.
    pub claim_owner: Option<u64>,
}

/// What an operation does, beyond answering its proposal.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum OperationKind {
    /// See [`OpKind::Noop`].
    Noop,
    /// See [`Register`].
    Register(Register),
    /// Remove one resource and its subtree.
    Unregister {
        /// Which collection it belongs to.
        resource_type: ResourceType,
        /// The resource's own id.
        resource_id: String,
    },
    /// A Node whose lease lapsed.
    Expire {
        /// The Node to remove, with its subtree.
        node_id: String,
    },
    /// Drop tombstones whose forget interval has elapsed.
    Forget {
        /// Exactly which records to drop, decided once by the proposer.
        victims: Vec<(ResourceType, String)>,
    },
    /// Take responsibility for a Node's subtree.
    ClaimOwnership {
        /// The Node being claimed.
        node_id: String,
        /// The member taking it.
        owner: u64,
    },
    /// Give it up.
    ReleaseOwnership {
        /// The Node being released.
        node_id: String,
    },
    /// One member is gone.
    MemberDown {
        /// Which member.
        member: u64,
    },
}

impl OperationKind {
    /// The kind byte this travels under.
    #[must_use]
    pub const fn kind(&self) -> OpKind {
        match *self {
            Self::Noop => OpKind::Noop,
            Self::Register(_) => OpKind::Register,
            Self::Unregister { .. } => OpKind::Unregister,
            Self::Expire { .. } => OpKind::Expire,
            Self::Forget { .. } => OpKind::Forget,
            Self::ClaimOwnership { .. } => OpKind::ClaimOwnership,
            Self::ReleaseOwnership { .. } => OpKind::ReleaseOwnership,
            Self::MemberDown { .. } => OpKind::MemberDown,
        }
    }
}

/// One replicated operation: who proposed it, and what it does.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Operation {
    /// Who proposed it, so the outcome can be returned to them.
    pub proposal: ProposalId,
    /// What it does.
    pub kind: OperationKind,
}

impl Operation {
    /// Serialise one operation. Field 1 is always the kind.
    #[must_use]
    pub fn encode(&self) -> Vec<u8> {
        let mut body = Writer::new().bytes(
            1,
            &Writer::new()
                .uint(1, self.proposal.member)
                .uint(2, self.proposal.sequence)
                .take(),
        );

        match self.kind {
            OperationKind::Noop => {}
            OperationKind::Register(ref op) => {
                body = body
                    .string(2, op.resource_type.singular())
                    .string(3, &op.resource_id)
                    .string(4, &op.node_id)
                    .string(5, &op.body_text);
                body = write_cursor(body, 6, op.created);
                body = write_cursor(body, 7, op.updated);
                body = body.uint(8, op.health).bool(9, op.expect_created);
                if let Some(owner) = op.claim_owner {
                    body = body.uint(10, owner);
                }
            }
            OperationKind::Unregister {
                resource_type,
                ref resource_id,
            } => {
                body = body
                    .string(2, resource_type.singular())
                    .string(3, resource_id);
            }
            OperationKind::Expire { ref node_id }
            | OperationKind::ReleaseOwnership { ref node_id } => {
                body = body.string(2, node_id);
            }
            OperationKind::Forget { ref victims } => {
                for &(resource_type, ref resource_id) in victims {
                    let victim = Writer::new()
                        .string(1, resource_type.singular())
                        .string(2, resource_id)
                        .take();
                    body = body.bytes(2, &victim);
                }
            }
            OperationKind::ClaimOwnership { ref node_id, owner } => {
                body = body.string(2, node_id).uint(3, owner);
            }
            OperationKind::MemberDown { member } => {
                body = body.uint(2, member);
            }
        }

        Writer::new()
            .uint(1, self.kind.kind() as u64)
            .bytes(2, &body.take())
            .take()
    }

    /// Parse one operation.
    ///
    /// Called on receipt rather than at apply time, deliberately. A malformed
    /// entry discovered here can still drop the link; discovered inside apply
    /// it would be a synchronous mutation that has nowhere to fail.
    ///
    /// # Errors
    ///
    /// [`RaftProtocolError`] if the payload is malformed, names no kind, or
    /// names a kind or resource type this build does not know.
    pub fn decode(data: &[u8]) -> Result<Self, RaftProtocolError> {
        let mut kind_value: Option<u64> = None;
        let mut body: &[u8] = &[];
        let mut reader = Reader::new(data);
        while let Some((number, wire)) = reader.next_field()? {
            match number {
                1 => kind_value = Some(reader.uint()?),
                2 => body = reader.bytes()?,
                _ => reader.skip(wire)?,
            }
        }

        let Some(kind_value) = kind_value else {
            return Err(RaftProtocolError("operation has no kind".to_owned()));
        };
        let Some(kind) = OpKind::from_byte(kind_value) else {
            return Err(RaftProtocolError(format!(
                "unknown operation kind {kind_value}"
            )));
        };

        let mut fields = Fields::default();
        let mut inner = Reader::new(body);
        while let Some((number, wire)) = inner.next_field()? {
            if number == 1 {
                fields.proposal = read_proposal(inner.bytes()?)?;
                continue;
            }
            fields.read(kind, number, wire, &mut inner)?;
        }

        Ok(Self {
            proposal: fields.proposal,
            kind: fields.into_kind(kind)?,
        })
    }
}

/// Everything a body can carry, before it is narrowed to one kind.
///
/// Flat, like the Python's local variables, because the fields a kind uses are
/// decided by the kind byte and the reader cannot know them until it has one.
struct Fields {
    proposal: ProposalId,
    resource_type_name: String,
    resource_id: String,
    node_id: String,
    body_text: String,
    created: TaiCursor,
    updated: TaiCursor,
    health: u64,
    expect_created: bool,
    claim_owner: Option<u64>,
    member: u64,
    victims: Vec<(ResourceType, String)>,
}

impl Default for Fields {
    fn default() -> Self {
        Self {
            proposal: ProposalId {
                member: 0,
                sequence: 0,
            },
            resource_type_name: String::new(),
            resource_id: String::new(),
            node_id: String::new(),
            body_text: String::new(),
            created: TaiCursor::MIN,
            updated: TaiCursor::MIN,
            health: 0,
            expect_created: false,
            claim_owner: None,
            member: 0,
            victims: Vec::new(),
        }
    }
}

impl Fields {
    fn read(
        &mut self,
        kind: OpKind,
        number: u64,
        wire: crate::wire::WireType,
        inner: &mut Reader<'_>,
    ) -> Result<(), RaftProtocolError> {
        match (kind, number) {
            (OpKind::Register, 2) => self.resource_type_name = inner.string()?,
            (OpKind::Register, 3) => self.resource_id = inner.string()?,
            (OpKind::Register, 4) => self.node_id = inner.string()?,
            (OpKind::Register, 5) => self.body_text = inner.string()?,
            (OpKind::Register, 6) => self.created = read_cursor(inner.bytes()?)?,
            (OpKind::Register, 7) => self.updated = read_cursor(inner.bytes()?)?,
            (OpKind::Register, 8) => self.health = inner.uint()?,
            (OpKind::Register, 9) => self.expect_created = inner.bool()?,
            (OpKind::Register, 10) => self.claim_owner = Some(inner.uint()?),

            (OpKind::Unregister, 2) => self.resource_type_name = inner.string()?,
            (OpKind::Unregister, 3) => self.resource_id = inner.string()?,

            (OpKind::Expire | OpKind::ReleaseOwnership | OpKind::ClaimOwnership, 2) => {
                self.node_id = inner.string()?;
            }
            (OpKind::ClaimOwnership, 3) => self.claim_owner = Some(inner.uint()?),

            (OpKind::Forget, 2) => {
                let victim = read_victim(inner.bytes()?)?;
                self.victims.push(victim);
            }

            (OpKind::MemberDown, 2) => self.member = inner.uint()?,

            _ => inner.skip(wire)?,
        }
        Ok(())
    }

    fn into_kind(self, kind: OpKind) -> Result<OperationKind, RaftProtocolError> {
        Ok(match kind {
            OpKind::Noop => OperationKind::Noop,
            OpKind::Register => OperationKind::Register(Register {
                resource_type: resource_type(&self.resource_type_name)?,
                resource_id: self.resource_id,
                node_id: self.node_id,
                body_text: self.body_text,
                created: self.created,
                updated: self.updated,
                health: self.health,
                expect_created: self.expect_created,
                claim_owner: self.claim_owner,
            }),
            OpKind::Unregister => OperationKind::Unregister {
                resource_type: resource_type(&self.resource_type_name)?,
                resource_id: self.resource_id,
            },
            OpKind::Expire => OperationKind::Expire {
                node_id: self.node_id,
            },
            OpKind::Forget => OperationKind::Forget {
                victims: self.victims,
            },
            OpKind::ClaimOwnership => {
                let Some(owner) = self.claim_owner else {
                    return Err(RaftProtocolError(
                        "ownership claim names no owner".to_owned(),
                    ));
                };
                OperationKind::ClaimOwnership {
                    node_id: self.node_id,
                    owner,
                }
            }
            OpKind::ReleaseOwnership => OperationKind::ReleaseOwnership {
                node_id: self.node_id,
            },
            OpKind::MemberDown => OperationKind::MemberDown {
                member: self.member,
            },
        })
    }
}

fn write_cursor(writer: Writer, field: u64, cursor: TaiCursor) -> Writer {
    let encoded = Writer::new()
        .uint(1, cursor.seconds)
        .uint(2, cursor.nanoseconds)
        .take();
    writer.bytes(field, &encoded)
}

fn read_cursor(payload: &[u8]) -> Result<TaiCursor, RaftProtocolError> {
    let mut seconds = 0;
    let mut nanoseconds = 0;
    let mut reader = Reader::new(payload);
    while let Some((number, wire)) = reader.next_field()? {
        match number {
            1 => seconds = reader.uint()?,
            2 => nanoseconds = reader.uint()?,
            _ => reader.skip(wire)?,
        }
    }
    // The struct literal, not `TaiCursor::new`: `new` normalises a nanosecond
    // field at or above one second into the seconds, and the Python's
    // `TaiCursor(seconds, nanoseconds)` does not. A cursor that arrived as
    // `0:5000000000` must stay `0:5000000000`, or two members would hold
    // different cursors for the same resource and page differently.
    Ok(TaiCursor {
        seconds,
        nanoseconds,
    })
}

fn read_proposal(payload: &[u8]) -> Result<ProposalId, RaftProtocolError> {
    let mut member = 0;
    let mut sequence = 0;
    let mut reader = Reader::new(payload);
    while let Some((number, wire)) = reader.next_field()? {
        match number {
            1 => member = reader.uint()?,
            2 => sequence = reader.uint()?,
            _ => reader.skip(wire)?,
        }
    }
    Ok(ProposalId { member, sequence })
}

fn read_victim(payload: &[u8]) -> Result<(ResourceType, String), RaftProtocolError> {
    let mut name = String::new();
    let mut resource_id = String::new();
    let mut reader = Reader::new(payload);
    while let Some((number, wire)) = reader.next_field()? {
        match number {
            1 => name = reader.string()?,
            2 => resource_id = reader.string()?,
            _ => reader.skip(wire)?,
        }
    }
    Ok((resource_type(&name)?, resource_id))
}

fn resource_type(value: &str) -> Result<ResourceType, RaftProtocolError> {
    ResourceType::from_singular(value)
        .ok_or_else(|| RaftProtocolError(format!("unknown resource type '{value}'")))
}
